flink-clients/src/main/java/org/apache/flink/client/program/OptimizerPlanEnvironment.java

```diff
@@ -22,143 +22,39 @@
 import org.apache.flink.api.java.ExecutionEnvironment;
 import org.apache.flink.api.java.ExecutionEnvironmentFactory;
 import org.apache.flink.core.execution.JobClient;
-import org.apache.flink.util.Preconditions;
-
-import javax.annotation.Nullable;
-
-import java.io.ByteArrayOutputStream;
-import java.io.PrintStream;
 
 /**
- * An {@link ExecutionEnvironment} that never executes a job but only extracts the {@link
- * org.apache.flink.api.dag.Pipeline}.
+ * An {@link ExecutionEnvironment} that never executes a job but only extracts the {@link Pipeline}.
  */
 public class OptimizerPlanEnvironment extends ExecutionEnvironment {
 
-	private Pipeline pipeline;
+	private static Pipeline pipeline;
```

Comment:
This change has some very subtle implications. Before, `OptimizerPlanEnvironment` was thread save, because `initializeContextEnvironment()` would store the environment factory in a thread-local field and the `Pipeline` field was not static. Now, multiple threads would try and store the pipeline in this one shared static field.

Reply:
What if users write code

```java
ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

parallel-n {
  env.execute();
}
```

In this case, the whole function is not thread transparent. And even with master branch we might miss the exception. Here, already we make some assumption that user calls `env.execute` in the same thread he runs `main` and doesn't caught `ProgramAbortException`.

If we keep `OptimizerPlanEnvironment` as is, then we go into the problem you mention above for ContextEnvironment. What if

```java
ExecutionEnvironment env1 = ExecutionEnvironment.getExecutionEnvironment();
ExecutionEnvironment env2 = ExecutionEnvironment.getExecutionEnvironment();

env1 ...
env2 ...

env1.execute()
```

Actually `env1` & `env2` is the same environment.

Either way, we make some implications. And if you compare master `OptimizerPlanEnvironment` and `StreamPlanEnvironment`, you will find that they are different.

- `OptimizerPlanEnvironment.setAsContext` always set the same instance
- `StreamPlanEnvironment.setAsContext` always set a new instance, though it finally set StreamGraph to the single `OptimizerPlanEnvironment`, but different `StreamPlanEnvironment` instance has different configuration and the first one call `execute()` wins. If you put it in multi-threaded environment, the one runs as the same thread of `main` and first call `execute()` wins.

Reply:
It is, unfortunately, a historical problem that becomes obvious in this pull request. And anyway, we can tweak the code a bit to keep the manner as is. For the specific case, we use single `OptimizerPlanEnvironment` and `ContextEnvironmentFactory` like `StreamPlanEnvironment`.